Mobx-utils/fromPromise.state is not Observable?

I try to react when `state` change.. but looks like `state` is not a observable...

    const result = fromPromise(promise);
    console.assert(isObservable(result.state), 'state is NOT an isObservable'); // AssertionError: state is NOT an isObservable

Full code.. 
https://runkit.com/ridermansb/mobx-react-is-observable/0.0.5

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,8 @@
 # 3.0.0
 
 * Revamped `fromPromise`:
+  * It is now possible to directly pass a `(resolve, reject) => {}` function to `fromPromise`, instead of a promise object
+  * **BREAKING** `fromPromise` no longer creates a wrapping object, but rather extends the given promise, #45
   * **BREAKING** Fixed #54, the resolved value of a promise is no longer deeply converted to an observable
   * **BREAKING** Dropped `fromPromise().reason`
   * **BREAKING** Improved typings of `fromPromise`. For example, the `value` property is now only available if `.state === "resolved"` (#41)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,11 +1,12 @@
 # 3.0.0
 
-* **BREAKING** Fixed #54, the resolved value of a promise is no longer deeply converted to an observable
-* **BREAKING** Dropped `fromPromise().reason`
-* **BREAKING** Improved typings of `fromPromise`. For example, the `value` property is now only available if `.state === "resolved"`
-* **BREAKING** Dropped optional `initialvalue` param from `fromPromise`. use `fromPromise.fullfilled(value)` instead to create a promise in some ready state (see #39)
-* Introduced `fromPromise.reject(reason)` and `fromPromise.resolve(value?)` to create a promise based observable in a certain state
-* Fixed #56, observable promises attributes `state` and `value` are now explicit observables
+* Revamped `fromPromise`:
+  * **BREAKING** Fixed #54, the resolved value of a promise is no longer deeply converted to an observable
+  * **BREAKING** Dropped `fromPromise().reason`
+  * **BREAKING** Improved typings of `fromPromise`. For example, the `value` property is now only available if `.state === "resolved"` (#41)
+  * **BREAKING** Dropped optional `initialvalue` param from `fromPromise`. use `fromPromise.fullfilled(value)` instead to create a promise in some ready state
+  * Introduced `fromPromise.reject(reason)` and `fromPromise.resolve(value?)` to create a promise based observable in a certain state, see #39
+  * Fixed #56, observable promises attributes `state` and `value` are now explicit observables
 
 # 2.0.2
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+# 2.0.3
+
+* Fixed #56, observable promises attributes `state` and `value` are now explicit observables
+
 # 2.0.2
 
 * Fixed #44, lazyObservable not accepting an array as initial value.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@
 * **BREAKING** Dropped `fromPromise().reason`
 * **BREAKING** Improved typings of `fromPromise`. For example, the `value` property is now only available if `.state === "resolved"`
 * **BREAKING** Dropped optional `initialvalue` param from `fromPromise`. use `fromPromise.fullfilled(value)` instead to create a promise in some ready state (see #39)
+* Introduced `fromPromise.reject(reason)` and `fromPromise.resolve(value?)` to create a promise based observable in a certain state
 * Fixed #56, observable promises attributes `state` and `value` are now explicit observables
 
 # 2.0.2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
-# 2.0.3
+# 3.0.0
 
+* **BREAKING** Fixed #54, the resolved value of a promise is no longer deeply converted to an observable
+* **BREAKING** Dropped `fromPromise().reason`
+* **BREAKING** Improved typings of `fromPromise`. For example, the `value` property is now only available if `.state === "resolved"`
+* **BREAKING** Dropped optional `initialvalue` param from `fromPromise`. use `fromPromise.fullfilled(value)` instead to create a promise in some ready state (see #39)
 * Fixed #56, observable promises attributes `state` and `value` are now explicit observables
 
 # 2.0.2

--- a/README.md
+++ b/README.md
@@ -24,8 +24,6 @@ CDN: <https://unpkg.com/mobx-utils/mobx-utils.umd.js>
 
 ## fromPromise
 
-[lib/from-promise.js:100-103](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/from-promise.js#L100-L103 "Source code on GitHub")
-
 `fromPromise` takes a Promise and returns an object with 3 observable properties that track
 the status of the promise. The returned object has the following observable properties:
 
@@ -79,8 +77,6 @@ Returns **IPromiseBasedObservable&lt;T>**
 
 ## isPromiseBasedObservable
 
-[lib/from-promise.js:110-112](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/from-promise.js#L110-L112 "Source code on GitHub")
-
 Returns true if the provided value is a promise-based observable.
 
 **Parameters**
@@ -90,8 +86,6 @@ Returns true if the provided value is a promise-based observable.
 Returns **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** 
 
 ## lazyObservable
-
-[lib/lazy-observable.js:37-72](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/lazy-observable.js#L37-L72 "Source code on GitHub")
 
 `lazyObservable` creates an observable around a `fetch` method that will not be invoked
 until the observable is needed the first time.
@@ -127,8 +121,6 @@ userProfile.refresh()
 ```
 
 ## fromResource
-
-[lib/from-resource.js:68-104](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/from-resource.js#L68-L104 "Source code on GitHub")
 
 `fromResource` creates an observable whose current state can be inspected using `.current()`,
 and which can be kept in sync with some external datasource that can be subscribed to.
@@ -193,8 +185,6 @@ const userComponent = observer(({ user }) =>
 
 ## toStream
 
-[lib/observable-stream.js:37-57](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/observable-stream.js#L37-L57 "Source code on GitHub")
-
 Converts an expression to an observable stream (a.k.a. TC 39 Observable / RxJS observable).
 The provided expression is tracked by mobx as long as there are subscribers, automatically
 emitting when new values become available. The expressions respect (trans)actions.
@@ -221,8 +211,6 @@ Returns **IObservableStream&lt;T>**
 
 ## fromStream
 
-[lib/observable-stream.js:122-125](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/observable-stream.js#L122-L125 "Source code on GitHub")
-
 Converts an subscribable, observable stream (TC 39 observable / RxJS stream)
 into an object which stores the current value (as `current`). The subscription can be cancelled through the `dispose` method.
 Takes an initial value as second optional argument
@@ -247,8 +235,6 @@ autorun(() => {
 ```
 
 ## createViewModel
-
-[lib/create-view-model.js:127-129](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/create-view-model.js#L127-L129 "Source code on GitHub")
 
 `createViewModel` takes an object with observable properties (model)
 and wraps a viewmodel around it. The viewmodel proxies all enumerable property of the original model with the following behavior:
@@ -277,7 +263,7 @@ Note that if you read a non-dirty property, viewmodel only proxies the read to t
 
 ```javascript
 class Todo {
-  @observable title = "Test"
+  \@observable title = "Test"
 }
 
 const model = new Todo()
@@ -298,8 +284,6 @@ viewModel.reset()
 ```
 
 ## whenWithTimeout
-
-[lib/guarded-when.js:32-51](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/guarded-when.js#L32-L51 "Source code on GitHub")
 
 Like normal `when`, except that this `when` will automatically dispose if the condition isn't met within a certain amount of time.
 
@@ -335,8 +319,6 @@ Returns **IDisposer** disposer function that can be used to cancel the when prem
 
 ## keepAlive
 
-[lib/keep-alive.js:31-36](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/keep-alive.js#L31-L36 "Source code on GitHub")
-
 **Parameters**
 
 -   `computedValue` **IComputedValue&lt;any>** created using the `computed` function
@@ -357,8 +339,6 @@ stop()
 Returns **IDisposer** stops this keep alive so that the computed value goes back to normal behavior
 
 ## keepAlive
-
-[lib/keep-alive.js:31-36](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/keep-alive.js#L31-L36 "Source code on GitHub")
 
 MobX normally suspends any computed value that is not in use by any reaction,
 and lazily re-evaluates the expression if needed outside a reaction while not in use.
@@ -385,8 +365,6 @@ Returns **IDisposer** stops this keep alive so that the computed value goes back
 
 ## queueProcessor
 
-[lib/queue-processor.js:22-40](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/queue-processor.js#L22-L40 "Source code on GitHub")
-
 `queueProcessor` takes an observable array, observes it and calls `processor`
 once for each item added to the observable array, optionally deboucing the action
 
@@ -412,8 +390,6 @@ pendingNotifications.push("test!")
 Returns **IDisposer** stops the processor
 
 ## chunkProcessor
-
-[lib/chunk-processor.js:27-52](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/chunk-processor.js#L27-L52 "Source code on GitHub")
 
 `chunkProcessor` takes an observable array, observes it and calls `processor`
 once for a chunk of items added to the observable array, optionally deboucing the action.
@@ -445,8 +421,6 @@ trackedActions.push("hoveredButton")
 Returns **IDisposer** stops the processor
 
 ## now
-
-[lib/now.js:30-39](https://github.com/mobxjs/mobx-utils/blob/2e0d881a4901917426571319d529adbc9a3b3afa/lib/now.js#L30-L39 "Source code on GitHub")
 
 Returns the current date time as epoch number.
 The date time is read from an observable which is updated automatically after the given interval.

--- a/package.json
+++ b/package.json
@@ -7,7 +7,7 @@
   "scripts": {
     "build": "tsc -p src && npm run webpack",
     "webpack": "webpack -p",
-    "test": "npm run build && tape test/*.js | faucet",
+    "test": "npm run build && tape test/*.js | faucet && tsc -p test",
     "lint:js": "eslint ./test",
     "lint:ts": "tslint ./src/*.ts",
     "lint": "npm run lint:ts && npm run lint:js",

--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
     "prepublish": "npm run build && npm run build-docs",
     "test-travis": "npm run build && npm run lint && istanbul cover tape test/*.js",
     "coverage": "npm run build && istanbul cover tape test/*.js && cat ./coverage/lcov.info|coveralls",
-    "build-docs": "npm run build && documentation readme lib/mobx-utils.js --github --section API"
+    "build-docs": "npm run build && documentation readme lib/mobx-utils.js --section API"
   },
   "repository": {
     "type": "git",

--- a/src/from-promise.ts
+++ b/src/from-promise.ts
@@ -1,5 +1,5 @@
 import {IObservableValue, observable, action} from "mobx";
-import {IDENTITY, deprecated} from "./utils";
+import {IDENTITY, deprecated, invariant} from "./utils";
 
 export type PromiseState = "pending" | "fulfilled" | "rejected";
 
@@ -7,38 +7,43 @@ export const PENDING = "pending";
 export const FULFILLED = "fulfilled";
 export const REJECTED = "rejected";
 
-export interface IPromiseBasedObservable<T> {
-    readonly value: T;
-    readonly state: PromiseState;
-    readonly reason: any;
+export type IBasePromiseBasedObservable<T> = {
     readonly promise: PromiseLike<T>;
     case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U;
 }
 
-class PromiseBasedObservable<T> implements IPromiseBasedObservable<T> {
+export type IPendingPromise = {
+    readonly state: "pending";
+    readonly reason: any;
+}
+
+export type IResolvedPromise<T> = {
+    readonly state: "fulfilled";
+    readonly value: T;
+}
+
+export type IRejectedPromise = {
+    readonly state: "rejected";
+    readonly value: any;
+}
+
+export type IPromiseBasedObservable<T> = IBasePromiseBasedObservable<T> & (IPendingPromise | IResolvedPromise<T> | IRejectedPromise)
+
+class PromiseBasedObservable<T> {
     @observable.ref value: T = undefined;
     @observable.ref state: PromiseState = PENDING;
-    @observable.ref _reason: any = undefined;
 
-    constructor(public promise: PromiseLike<T>, initialValue: T = undefined) {
-        this.value = initialValue;
+    constructor(public promise: PromiseLike<T>) {
         promise.then(
             action("observableFromPromise-resolve", (value: T) => {
                 this.value = value;
                 this.state = FULFILLED;
             }),
             action("observableFromPromise-reject", (reason: any) => {
-                this._reason = reason;
                 this.value = reason;
                 this.state = REJECTED;
             })
         );
-    }
-
-    get reason(): any {
-        // TODO: kill in next major
-        deprecated("In `fromPromise`: `.reason` is deprecated, use `.value` instead");
-        return this._reason;
     }
 
     public case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U {
@@ -95,8 +100,10 @@ class PromiseBasedObservable<T> implements IPromiseBasedObservable<T> {
  * @param {T} [initialValue=undefined] Optional predefined initial value
  * @returns {IPromiseBasedObservable<T>}
  */
-export function fromPromise<T>(promise: PromiseLike<T>, initialValue: T = undefined): IPromiseBasedObservable<T> {
-    return new PromiseBasedObservable(promise, initialValue);
+export function fromPromise<T>(promise: PromiseLike<T>): IPromiseBasedObservable<T> {
+    invariant(arguments.length === 1, "fromPromise expects exactly one argument");
+    invariant(typeof promise === "object" && promise && typeof promise.then === "function", "Please pass a promise to fromPromise");
+    return new PromiseBasedObservable(promise) as any;
 }
 
 /**

--- a/src/from-promise.ts
+++ b/src/from-promise.ts
@@ -8,49 +8,44 @@ export const FULFILLED = "fulfilled";
 export const REJECTED = "rejected";
 
 export interface IPromiseBasedObservable<T> {
-    value: T;
-    state: PromiseState;
-    reason: any;
-    promise: PromiseLike<T>;
+    readonly value: T;
+    readonly state: PromiseState;
+    readonly reason: any;
+    readonly promise: PromiseLike<T>;
     case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U;
 }
 
 class PromiseBasedObservable<T> implements IPromiseBasedObservable<T> {
-    private _observable: IObservableValue<T>;
-    private _state: IObservableValue<PromiseState> = observable(PENDING as any); // MWE: Hm... as any should not be needed...
-    private _reason: IObservableValue<any> = observable.shallowBox(undefined as any);
+    @observable.ref value: T = undefined;
+    @observable.ref state: PromiseState = PENDING;
+    @observable.ref _reason: any = undefined;
 
     constructor(public promise: PromiseLike<T>, initialValue: T = undefined) {
-        this._observable = observable.box(initialValue);
+        this.value = initialValue;
         promise.then(
             action("observableFromPromise-resolve", (value: T) => {
-                this._observable.set(value);
-                this._state.set("fulfilled");
+                this.value = value;
+                this.state = FULFILLED;
             }),
             action("observableFromPromise-reject", (reason: any) => {
-                this._reason.set(reason);
-                this._observable.set(reason);
-                this._state.set("rejected");
+                this._reason = reason;
+                this.value = reason;
+                this.state = REJECTED;
             })
         );
     }
 
-    get value(): T {
-        return this._observable.get();
-    }
-    get state(): PromiseState {
-        return this._state.get();
-    }
     get reason(): any {
+        // TODO: kill in next major
         deprecated("In `fromPromise`: `.reason` is deprecated, use `.value` instead");
-        return this._reason.get();
+        return this._reason;
     }
 
     public case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U {
         switch (this.state) {
-            case "pending": return handlers.pending && handlers.pending();
-            case "rejected": return handlers.rejected && handlers.rejected(this.value);
-            case "fulfilled": return handlers.fulfilled && handlers.fulfilled(this.value);
+            case PENDING: return handlers.pending && handlers.pending();
+            case REJECTED: return handlers.rejected && handlers.rejected(this.value);
+            case FULFILLED: return handlers.fulfilled && handlers.fulfilled(this.value);
         }
     }
 }

--- a/src/from-promise.ts
+++ b/src/from-promise.ts
@@ -1,4 +1,4 @@
-import {IObservableValue, observable, action} from "mobx";
+import {IObservableValue, extendShallowObservable, action} from "mobx";
 import {IDENTITY, deprecated, invariant} from "./utils";
 
 export type PromiseState = "pending" | "fulfilled" | "rejected";
@@ -8,9 +8,9 @@ export const FULFILLED = "fulfilled";
 export const REJECTED = "rejected";
 
 export type IBasePromiseBasedObservable<T> = {
-    readonly promise: PromiseLike<T>;
+    isPromiseBasedObservable: true,
     case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U;
-}
+} & PromiseLike<T>
 
 export type IPendingPromise = {
     readonly state: "pending";
@@ -29,30 +29,56 @@ export type IRejectedPromise = {
 
 export type IPromiseBasedObservable<T> = IBasePromiseBasedObservable<T> & (IPendingPromise | IFulfilledPromise<T> | IRejectedPromise)
 
-class PromiseBasedObservable<T> {
-    @observable.ref value: T = undefined;
-    @observable.ref state: PromiseState = PENDING;
+function caseImpl<U, T>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U {
+    switch (this.state) {
+        case PENDING: return handlers.pending && handlers.pending();
+        case REJECTED: return handlers.rejected && handlers.rejected(this.value);
+        case FULFILLED: return handlers.fulfilled && handlers.fulfilled(this.value);
+    }
+}
 
-    constructor(public promise: PromiseLike<T>) {
-        promise.then(
-            action("observableFromPromise-resolve", (value: T) => {
-                this.value = value;
-                this.state = FULFILLED;
-            }),
-            action("observableFromPromise-reject", (reason: any) => {
-                this.value = reason;
-                this.state = REJECTED;
-            })
-        );
+function createObservablePromise(origPromise: any) {
+    invariant(arguments.length === 1, "fromPromise expects exactly one argument");
+    invariant(
+        typeof origPromise === "function" ||
+            (typeof origPromise === "object" && origPromise && typeof origPromise.then === "function"),
+        "Please pass a promise or function to fromPromise"
+    );
+
+    if (typeof origPromise === "function") {
+        // If it is a (reject, resolve function, wrap it)
+        origPromise = new Promise(origPromise as any);
     }
 
-    public case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U {
-        switch (this.state) {
-            case PENDING: return handlers.pending && handlers.pending();
-            case REJECTED: return handlers.rejected && handlers.rejected(this.value);
-            case FULFILLED: return handlers.fulfilled && handlers.fulfilled(this.value);
+    const promise = new Promise((resolve, reject) => {
+        origPromise.then(action("observableFromPromise-resolve", (value: any) => {
+            promise.value = value;
+            promise.state = FULFILLED;
+            resolve(value);
+        }),
+        action("observableFromPromise-reject", (reason: any) => {
+            promise.value = reason;
+            promise.state = REJECTED;
+            reject(reason);
+        }));
+    }) as any;
+
+    promise.isPromiseBasedObservable = true;
+    promise.case = caseImpl;
+    extendShallowObservable(promise, {
+        value: undefined,
+        state: PENDING
+    });
+
+    // TODO: remove in next major
+    Object.defineProperty(promise, "promise", {
+        get() {
+            deprecated("fromPromise().promise is deprecated. fromPromise now directly returns a promise");
+            return origPromise;
         }
-    }
+    });
+
+    return promise;
 }
 
 /**
@@ -64,7 +90,19 @@ class PromiseBasedObservable<T> {
  * and the following method:
  * - `case({fulfilled, rejected, pending})`: maps over the result using the provided handlers, or returns `undefined` if a handler isn't available for the current promise state.
  *
+ *
+ * Note that the status strings are available as constants:
+ * `mobxUtils.PENDING`, `mobxUtils.REJECTED`, `mobxUtil.FULFILLED`
+ *
+ * Observable promises can be created immediatly in a certain state using
+ * `fromPromise.reject(reason)` or `fromPromise.resolve(value?)`.
+ * The mean advantagate of `fromPromise.resolve(value)` over `fromPromise(Promise.resolve(value))` is that the first _synchronously_ starts in the desired state.
+ *
+ * It is possible to directly create a promise using a resolve, reject function:
+ * `fromPromise((resolve, reject) => setTimeout(() => resolve(true), 1000))`
+ *
  * @example
+ * ```javascript
  * const fetchResult = fromPromise(fetch("http://someurl"))
  *
  * // combine with when..
@@ -92,13 +130,7 @@ class PromiseBasedObservable<T> {
  *     rejected:  error => <div>Ooops.. {error}</div>
  *     fulfilled: value => <div>Gotcha: {value}</div>
  *   }))
- *
- * Note that the status strings are available as constants:
- * `mobxUtils.PENDING`, `mobxUtils.REJECTED`, `mobxUtil.FULFILLED`
- *
- * Observable promises can be created immediatly in a certain state using
- * `fromPromise.reject(reason)` or `fromPromise.resolve(value?)`.
- * The mean advantagate of `fromPromise.resolve(value)` over `fromPromise(Promise.resolve(value))` is that the first _synchronously_ starts in the desired state.
+ * ```
  *
  * @param {IThenable<T>} promise The promise which will be observed
  * @returns {IPromiseBasedObservable<T>}
@@ -107,11 +139,7 @@ export const fromPromise: {
     <T>(promise: PromiseLike<T>): IPromiseBasedObservable<T>;
     reject<T>(reason: any): IRejectedPromise & IBasePromiseBasedObservable<T>;
     resolve<T>(value?: T): IFulfilledPromise<T> & IBasePromiseBasedObservable<T>;
-} = function(promise: any) {
-    invariant(arguments.length === 1, "fromPromise expects exactly one argument");
-    invariant(typeof promise === "object" && promise && typeof promise.then === "function", "Please pass a promise to fromPromise");
-    return new PromiseBasedObservable(promise) as any;
-} as any;
+} = createObservablePromise as any;
 
 fromPromise.reject = action("fromPromise.reject", function(reason: any) {
     const p: any = fromPromise(Promise.reject(reason));
@@ -133,5 +161,5 @@ fromPromise.resolve = action("fromPromise.resolve", function(value: any = undefi
  * @returns {boolean}
  */
 export function isPromiseBasedObservable(value: any): value is IPromiseBasedObservable<any> {
-    return value instanceof PromiseBasedObservable;
+    return value && value.isPromiseBasedObservable === true;
 }

--- a/src/from-promise.ts
+++ b/src/from-promise.ts
@@ -96,11 +96,11 @@ class PromiseBasedObservable<T> {
  * Note that the status strings are available as constants:
  * `mobxUtils.PENDING`, `mobxUtils.REJECTED`, `mobxUtil.FULFILLED`
  *
- * For testing, promises can be created immediatly in a certain state using
- * `fromPromise.reject(reason)` or `fromPromise.resolve(reason)`
+ * Observable promises can be created immediatly in a certain state using
+ * `fromPromise.reject(reason)` or `fromPromise.resolve(value?)`.
+ * The mean advantagate of `fromPromise.resolve(value)` over `fromPromise(Promise.resolve(value))` is that the first _synchronously_ starts in the desired state.
  *
  * @param {IThenable<T>} promise The promise which will be observed
- * @param {T} [initialValue=undefined] Optional predefined initial value
  * @returns {IPromiseBasedObservable<T>}
  */
 export const fromPromise: {
@@ -128,10 +128,10 @@ fromPromise.resolve = action("fromPromise.resolve", function(value: any = undefi
 }) as any;
 
 /**
-  * Returns true if the provided value is a promise-based observable.
-  * @param value any
-  * @returns {boolean}
-  */
+ * Returns true if the provided value is a promise-based observable.
+ * @param value any
+ * @returns {boolean}
+ */
 export function isPromiseBasedObservable(value: any): value is IPromiseBasedObservable<any> {
     return value instanceof PromiseBasedObservable;
 }

--- a/src/from-promise.ts
+++ b/src/from-promise.ts
@@ -17,7 +17,7 @@ export type IPendingPromise = {
     readonly reason: any;
 }
 
-export type IResolvedPromise<T> = {
+export type IFulfilledPromise<T> = {
     readonly state: "fulfilled";
     readonly value: T;
 }
@@ -27,7 +27,7 @@ export type IRejectedPromise = {
     readonly value: any;
 }
 
-export type IPromiseBasedObservable<T> = IBasePromiseBasedObservable<T> & (IPendingPromise | IResolvedPromise<T> | IRejectedPromise)
+export type IPromiseBasedObservable<T> = IBasePromiseBasedObservable<T> & (IPendingPromise | IFulfilledPromise<T> | IRejectedPromise)
 
 class PromiseBasedObservable<T> {
     @observable.ref value: T = undefined;
@@ -96,15 +96,36 @@ class PromiseBasedObservable<T> {
  * Note that the status strings are available as constants:
  * `mobxUtils.PENDING`, `mobxUtils.REJECTED`, `mobxUtil.FULFILLED`
  *
+ * For testing, promises can be created immediatly in a certain state using
+ * `fromPromise.reject(reason)` or `fromPromise.resolve(reason)`
+ *
  * @param {IThenable<T>} promise The promise which will be observed
  * @param {T} [initialValue=undefined] Optional predefined initial value
  * @returns {IPromiseBasedObservable<T>}
  */
-export function fromPromise<T>(promise: PromiseLike<T>): IPromiseBasedObservable<T> {
+export const fromPromise: {
+    <T>(promise: PromiseLike<T>): IPromiseBasedObservable<T>;
+    reject<T>(reason: any): IRejectedPromise & IBasePromiseBasedObservable<T>;
+    resolve<T>(value?: T): IFulfilledPromise<T> & IBasePromiseBasedObservable<T>;
+} = function(promise: any) {
     invariant(arguments.length === 1, "fromPromise expects exactly one argument");
     invariant(typeof promise === "object" && promise && typeof promise.then === "function", "Please pass a promise to fromPromise");
     return new PromiseBasedObservable(promise) as any;
-}
+} as any;
+
+fromPromise.reject = action("fromPromise.reject", function(reason: any) {
+    const p: any = fromPromise(Promise.reject(reason));
+    p.state = REJECTED;
+    p.value = reason;
+    return p;
+}) as any;
+
+fromPromise.resolve = action("fromPromise.resolve", function(value: any = undefined) {
+    const p: any = fromPromise(Promise.resolve(value));
+    p.state = FULFILLED;
+    p.value = value;
+    return p;
+}) as any;
 
 /**
   * Returns true if the provided value is a promise-based observable.

--- a/src/tsconfig.json
+++ b/src/tsconfig.json
@@ -9,7 +9,10 @@
         "outDir": "../lib/",
         "noImplicitAny": true,
         "moduleResolution": "node",
-        "experimentalDecorators": true
+        "experimentalDecorators": true,
+        "lib": [
+            "dom", "es5", "scripthost", "es2015.promise"
+        ]
     },
     "exclude": [
         "node_modules"

--- a/test/from-promise.js
+++ b/test/from-promise.js
@@ -177,6 +177,54 @@ test('test from-promise', t => {
     })
   })
 
+  test('it is possible to create a promise in a rejected state, #36', t=> {
+    const someObject = { a: 3 }
+    const obs = utils.fromPromise.reject(someObject);
+    t.is(obs.state, utils.REJECTED);
+    t.is(obs.value, someObject);
+
+    // still a real promise backing it, which can be thenned...
+    obs.promise.catch((v) => {
+      t.is(obs.state, utils.REJECTED)
+      t.false(mobx.isObservable(obs.value))
+      t.true(obs.value === someObject)
+      t.true(v === someObject)
+      t.end()
+    })
+  })
+
+  test('it is possible to create a promise in a fullfilled state, #36', t=> {
+    const someObject = { a: 3 }
+    const obs = utils.fromPromise.resolve(someObject);
+    t.is(obs.state, utils.FULFILLED);
+    t.is(obs.value, someObject);
+
+    // still a real promise backing it, which can be thenned...
+    obs.promise.then((v) => {
+      t.is(obs.state, utils.FULFILLED)
+      t.false(mobx.isObservable(obs.value))
+      t.true(obs.value === someObject)
+      t.true(v === someObject)
+      t.end()
+    })
+  })
+
+  test('when creating a promise in a fullfilled state it should not fire twice, #36', t=> {
+    let events = 0;
+    const obs = utils.fromPromise.resolve(3);
+
+    mobx.autorun(() => {
+      obs.state; // track state & value
+      obs.value;
+      events++;
+    })
+
+    obs.promise.then((v) => {
+      t.is(events, 1); // only initial run should have run
+      t.end()
+    })
+  })
+
   t.end();
 });
 

--- a/test/from-promise.js
+++ b/test/from-promise.js
@@ -43,6 +43,20 @@ test('test from-promise', t => {
     );
   });
 
+test('resolves value from promise function', t => {
+    const obs = utils.fromPromise(resolve => resolve(7));
+    t.equal(obs.value, undefined);
+    t.equal(obs.state, 'pending');
+
+    mobx.when(
+      () => obs.value === 7,
+      () => {
+        t.equal(obs.state, utils.FULFILLED);
+        t.end();
+      }
+    );
+  });
+
   test('rejects with reason value', t => {
     const p = new Promise((resolve, reject) => {
       reject(7);
@@ -52,6 +66,23 @@ test('test from-promise', t => {
     t.equal(obs.value, undefined);
     t.equal(obs.state, 'pending');
     t.ok(obs.promise === p);
+
+    mobx.when(
+      () => obs.state !== utils.PENDING,
+      () => {
+        t.equal(obs.state, utils.REJECTED);
+        t.equal(obs.value, 7);
+        t.end();
+      }
+    );
+  });
+
+  test('rejects with reason value from fn', t => {
+    const obs = utils.fromPromise((resolve, reject) => {
+      reject(7);
+    });
+    t.equal(obs.value, undefined);
+    t.equal(obs.state, 'pending');
 
     mobx.when(
       () => obs.state !== utils.PENDING,
@@ -224,6 +255,29 @@ test('test from-promise', t => {
       t.end()
     })
   })
+
+  test('it creates a real promise, #45', t => {
+    Promise.all([
+      utils.fromPromise.resolve(2),
+      utils.fromPromise(Promise.resolve(3))
+    ]).then(x => {
+      t.deepEqual(x, [2, 3])
+      t.end()
+    })
+  })
+
+  test('it can construct new promises from function, #45', t => {
+    Promise.all([
+      utils.fromPromise((resolve, reject) => {
+        setTimeout(() => resolve(2), 200)
+      }),
+      utils.fromPromise(Promise.resolve(3))
+    ]).then(x => {
+      t.deepEqual(x, [2, 3])
+      t.end()
+    })
+  })
+
 
   t.end();
 });

--- a/test/from-promise.js
+++ b/test/from-promise.js
@@ -10,10 +10,9 @@ test('test from-promise', t => {
   test('resolves', t => {
     const p = new Promise(resolve => resolve(7));
 
-    const obs = utils.fromPromise(p, 3);
-    t.equal(obs.value, 3);
+    const obs = utils.fromPromise(p);
+    t.equal(obs.value, undefined);
     t.equal(obs.state, 'pending');
-    t.equal(obs.reason, undefined);
     t.ok(obs.promise === p);
 
     mobx.when(
@@ -30,10 +29,9 @@ test('test from-promise', t => {
   test('resolves value', t => {
     const p = new Promise(resolve => resolve(7));
 
-    const obs = utils.fromPromise(p, 3);
-    t.equal(obs.value, 3);
+    const obs = utils.fromPromise(p);
+    t.equal(obs.value, undefined);
     t.equal(obs.state, 'pending');
-    t.equal(obs.reason, undefined);
     t.ok(obs.promise === p);
 
     mobx.when(
@@ -50,10 +48,9 @@ test('test from-promise', t => {
       reject(7);
     });
 
-    const obs = utils.fromPromise(p, 3);
-    t.equal(obs.value, 3);
+    const obs = utils.fromPromise(p);
+    t.equal(obs.value, undefined);
     t.equal(obs.state, 'pending');
-    t.equal(obs.reason, undefined);
     t.ok(obs.promise === p);
 
     mobx.when(
@@ -61,7 +58,6 @@ test('test from-promise', t => {
       () => {
         t.equal(obs.state, utils.REJECTED);
         t.equal(obs.value, 7);
-        t.equal(obs.reason, 7);
         t.end();
       }
     );
@@ -72,10 +68,9 @@ test('test from-promise', t => {
       throw 7;
     });
 
-    const obs = utils.fromPromise(p, 3);
-    t.equal(obs.value, 3);
+    const obs = utils.fromPromise(p);
+    t.equal(obs.value, undefined);
     t.equal(obs.state, 'pending');
-    t.equal(obs.reason, undefined);
     t.ok(obs.promise === p);
 
     mobx.when(
@@ -83,7 +78,6 @@ test('test from-promise', t => {
       () => {
         t.equal(obs.state, 'rejected');
         t.equal(obs.value, 7);
-        t.equal(obs.reason, 7);
         t.end();
       }
     );
@@ -169,6 +163,18 @@ test('test from-promise', t => {
     t.ok(mobx.isObservable(obs, "state"))
     t.ok(mobx.isObservable(obs, "value"))
     t.end()
+  })
+
+  test('the resolved value of a promise is not convertd to some deep observable, #54', t => {
+    const someObject = { a: 3 }
+    const obs = utils.fromPromise(Promise.resolve(someObject));
+    obs.promise.then((v) => {
+      t.is(obs.state, utils.FULFILLED)
+      t.false(mobx.isObservable(obs.value))
+      t.true(obs.value === someObject)
+      t.true(v === someObject)
+      t.end()
+    })
   })
 
   t.end();

--- a/test/from-promise.js
+++ b/test/from-promise.js
@@ -151,18 +151,26 @@ test('test from-promise', t => {
       }
     );
   });
-  
+
   test('isPromiseBasedObservable, true', t => {
     const obs = utils.fromPromise(Promise.resolve(123));
     t.ok(utils.isPromiseBasedObservable(obs));
     t.end();
   });
-  
+
   test('isPromiseBasedObservable, false', t => {
     t.notOk(utils.isPromiseBasedObservable({}));
     t.end();
   });
-  
+
+  test('state and value are observable, #56', t => {
+    const obs = utils.fromPromise(Promise.resolve(123));
+    t.ok(mobx.isObservable(obs))
+    t.ok(mobx.isObservable(obs, "state"))
+    t.ok(mobx.isObservable(obs, "value"))
+    t.end()
+  })
+
   t.end();
 });
 

--- a/test/tsconfig.json
+++ b/test/tsconfig.json
@@ -1,0 +1,20 @@
+{
+    "version": "1.8.10",
+    "compilerOptions": {
+        "noEmit": true,
+        "target": "es5",
+        "module": "commonjs",
+        "noImplicitAny": true,
+        "moduleResolution": "node",
+        "experimentalDecorators": true,
+        "lib": [
+            "dom", "es5", "scripthost", "es2015.promise"
+        ]
+    },
+    "files": [
+        "type-tests.ts"
+    ],
+    "exclude": [
+        "../node_modules"
+    ]
+}

--- a/test/type-tests.ts
+++ b/test/type-tests.ts
@@ -9,3 +9,10 @@ import {fromPromise, FULFILLED} from "../src/mobx-utils";
         p.value.x = 4; // value only available if state is checked!
     }
 }
+
+{
+    // typings: can create a resolved promise
+    const x = { x: 3 };
+    const p = fromPromise.resolve(x);
+    p.value.x = 7;
+}

--- a/test/type-tests.ts
+++ b/test/type-tests.ts
@@ -1,0 +1,11 @@
+import {fromPromise, FULFILLED} from "../src/mobx-utils";
+
+{
+    // test typings of fromPromise
+    const x = { x: 3 };
+    const p = fromPromise(Promise.resolve(x));
+    // p.value // compile error!
+    if (p.state === FULFILLED) {
+        p.value.x = 4; // value only available if state is checked!
+    }
+}

